src/targeting/Semgrepignore.ml

```diff
@@ -120,14 +123,24 @@ let create ?include_patterns ?(cli_patterns = []) ~builtin_semgrepignore
   in
   let gitignore_filenames =
     match exclusion_mechanism with
-    | Gitignore_and_semgrepignore ->
-        [ ("gitignore", ".gitignore"); ("semgrepignore", ".semgrepignore") ]
-    | Only_semgrepignore -> [ ("semgrepignore", ".semgrepignore") ]
+    | Gitignore_and_semgrepignore -> [ gitignore_files; semgrep_ignore_files ]
+    | Only_semgrepignore -> [ semgrep_ignore_files ]
+  in
+  (* Check if there is a top level .semgrepignore. If not use builtins *)
+  let semgrep_ignore_exists =
+    let gitignore_cache =
+      Gitignores_cache.create ~gitignore_filenames:[ semgrep_ignore_files ]
+        ~project_root ()
+    in
+    Gitignores_cache.load gitignore_cache Ppath.root |> Option.is_some
+  in
+  let higher_priority_levels =
+    if semgrep_ignore_exists then [ cli_level ]
+    else [ builtin_level; cli_level ]
   in
   let gitignore_filter =
-    Gitignore_filter.create
-      ~higher_priority_levels:[ builtin_level; cli_level ]
-      ~gitignore_filenames ~project_root ()
+    Gitignore_filter.create ~higher_priority_levels ~gitignore_filenames
+      ~project_root ()
```

Comment:
We should rename `Gitignores_cache` to `Gitignore_cache`. @aryx gitignore cache means "cache for gitignores"; there's no "s" after the noun that's used as an adjective. Same for a "rule file" that contains multiple rules.

Reply:
Separate PR for this ok?